File: testing/tests_js/test_034_login_buttons.py
"""Node-backed checks for login action-button state."""


# @features login
# @dimensions submit-button loading-state
def test_login_action_button_uses_fixed_icon_and_text_slots(run_node):
    run_node(
        r"""
const fs = require("node:fs");
const vm = require("node:vm");

class FakeElement {
  constructor() {
    this.children = [];
    this.dataset = {};
    this.textContent = "";
    this.attributes = {};
  }

  querySelector(selector) {
    const match = selector.match(/\[data-role='([^']+)'\]/);
    return match
      ? this.children.find((child) => child.dataset.role === match[1]) || null
      : null;
  }

  replaceChildren(...children) {
    this.children = children;
  }

  prepend(child) {
    this.children.unshift(child);
  }

  setAttribute(name, value) {
    this.attributes[name] = value;
  }
}

const createIcon = (name) => {
  const icon = new FakeElement();
  icon.dataset.icon = name;
  return icon;
};
const context = {
  createIcon,
  document: {
    createElement: () => new FakeElement(),
  },
};
vm.createContext(context);
let source = fs.readFileSync("src/script/login/forms.mjs", "utf8");
source = source.replace(/import[\s\S]*?from ".*?";\n/g, "");
source = source.replace(
  /export \{[\s\S]*?\};/,
  "globalThis.setLoginActionButton = setLoginActionButton;",
);
vm.runInContext(source, context);

const button = new FakeElement();
const iconSlot = new FakeElement();
iconSlot.dataset.role = "icon";
iconSlot.dataset.visible = "false";
const textSlot = new FakeElement();
textSlot.dataset.role = "text";
textSlot.textContent = "Continue";
button.replaceChildren(iconSlot, textSlot);

context.setLoginActionButton(button, "Checking Email", "spinner");
if (button.children[0] !== iconSlot || button.children[1] !== textSlot) {
  throw new Error("Login button replaced its fixed icon or text slot");
}
if (
  iconSlot.dataset.visible !== "true" ||
  iconSlot.children[0]?.dataset.icon !== "spinner"
) {
  throw new Error("Login button did not render its spinner in the icon slot");
}
if (textSlot.textContent !== "Checking Email") {
  throw new Error("Login button did not update its text slot");
}

context.setLoginActionButton(button, "Continue");
if (iconSlot.dataset.visible !== "false" || iconSlot.children.length !== 0) {
  throw new Error("Login button did not clear and hide its icon slot");
}
if (textSlot.textContent !== "Continue") {
  throw new Error("Login button did not restore its text slot");
}

const unstructuredButton = new FakeElement();
context.setLoginActionButton(unstructuredButton, "Sign In", "spinner");
if (
  unstructuredButton.children[0]?.dataset.role !== "icon" ||
  unstructuredButton.children[1]?.dataset.role !== "text"
) {
  throw new Error("Login helper did not normalize an unstructured button");
}
"""
    )


# @features login
# @dimensions owner-bootstrap disabled-provider
def test_owner_setup_supports_password_only_mode(run_node):
    run_node(
        r"""
const fs = require("node:fs");
const vm = require("node:vm");

class FakeClassList {
  constructor(...names) { this.names = new Set(names); }
  add(name) { this.names.add(name); }
  remove(name) { this.names.delete(name); }
  contains(name) { return this.names.has(name); }
}

class FakeElement {
  constructor(selectors = {}, classes = []) {
    this.selectors = selectors;
    this.classList = new FakeClassList(...classes);
    this.dataset = {};
    this.textContent = "";
    this.value = "";
  }
  querySelector(selector) { return this.selectors[selector] || null; }
  querySelectorAll() { return []; }
  addEventListener() {}
  setAttribute(name, value) { this[name] = value; }
  replaceChildren(...children) { this.children = children; }
  prepend(child) { this.children = [child, ...(this.children || [])]; }
  focus() { this.focused = true; }
}

const googleError = new FakeElement({}, ["hidden"]);
const passwordError = new FakeElement({}, ["hidden"]);
const success = new FakeElement({}, ["hidden"]);
const password = new FakeElement();
const googleSetup = new FakeElement(
  { "[data-role='error']": googleError },
  ["hidden"],
);
const passwordSetup = new FakeElement({
  "[data-role='error']": passwordError,
  "input[type='password']": password,
});
const actionText = new FakeElement();
actionText.textContent = "Create Password";
actionText.dataset.role = "text";
const actionButton = new FakeElement({ "[data-role='text']": actionText });
const form = new FakeElement({
  "[data-role='error']": googleError,
  "[data-role='success']": success,
  "[data-role='owner-google-setup']": googleSetup,
  "[data-role='owner-password-setup']": passwordSetup,
  "[data-role='signin']": actionButton,
}, ["hidden"]);

const context = {
  createIcon() { return new FakeElement(); },
  document: { createElement: () => new FakeElement() },
};
vm.createContext(context);
let source = fs.readFileSync("src/script/login/forms.mjs", "utf8");
source = source.replace(/import[\s\S]*?from ".*?";\n/g, "");
source = source.replace(
  /export \{[\s\S]*?\};/,
  "globalThis.OwnerSetupForm = OwnerSetupForm;",
);
vm.runInContext(source, context);

const ownerSetup = new context.OwnerSetupForm({}, form);
ownerSetup.data = { email: "owner@example.test", error: "Provider unavailable" };
ownerSetup.show();

if (ownerSetup.error !== passwordError) {
  throw new Error("Password-only owner setup did not select its visible error slot");
}
if (passwordSetup.classList.contains("hidden")) {
  throw new Error("Password-only owner setup hid its available setup pane");
}
if (form.classList.contains("hidden")) {
  throw new Error("Password-only owner setup did not open");
}
if (passwordError.textContent !== "Provider unavailable") {
  throw new Error("Password-only owner setup did not display its error safely");
}
"""
    )

File: testing/tests_js/test_034_login_buttons.mjs
import assert from "node:assert/strict";
import { test } from "node:test";
import { createBrowser } from "../utility/js/environment.mjs";

/** @matrix login : loading-state submit-button */
test("test_login_action_button_uses_fixed_icon_and_text_slots", async (t) => {
	createBrowser(t, {
		html: `
			<button data-role="signin">
				<span data-role="icon" data-visible="false"></span>
				<span data-role="text">Continue</span>
			</button>
		`,
	});
	const { setLoginActionButton } = await import(
		"../../src/script/login/forms.mjs"
	);
	const button = document.querySelector("[data-role='signin']");
	const iconSlot = button.querySelector("[data-role='icon']");
	const textSlot = button.querySelector("[data-role='text']");

	setLoginActionButton(button, "Checking Email", "spinner");
	assert.equal(
		button.children[0],
		iconSlot,
		"Login button replaced its fixed icon slot",
	);
	assert.equal(
		button.children[1],
		textSlot,
		"Login button replaced its fixed text slot",
	);
	assert.equal(iconSlot.dataset.visible, "true");
	assert.equal(
		iconSlot.firstElementChild?.dataset.icon,
		"spinner",
		"Login button did not render its spinner in the icon slot",
	);
	assert.equal(
		textSlot.textContent,
		"Checking Email",
		"Login button did not update its text slot",
	);

	setLoginActionButton(button, "Continue");
	assert.equal(
		iconSlot.dataset.visible,
		"false",
		"Login button did not hide its icon slot",
	);
	assert.equal(
		iconSlot.children.length,
		0,
		"Login button did not clear its icon slot",
	);
	assert.equal(
		textSlot.textContent,
		"Continue",
		"Login button did not restore its text slot",
	);

	const unstructuredButton = document.createElement("button");
	setLoginActionButton(unstructuredButton, "Sign In", "spinner");
	assert.equal(
		unstructuredButton.children[0]?.dataset.role,
		"icon",
		"Login helper did not create an icon slot",
	);
	assert.equal(
		unstructuredButton.children[1]?.dataset.role,
		"text",
		"Login helper did not create a text slot",
	);
});

/** @matrix login : disabled-provider owner-bootstrap */
test("test_owner_setup_supports_password_only_mode", async (t) => {
	createBrowser(t, {
		html: `
			<form data-role="owner-setup" class="hidden">
				<p data-role="error" class="hidden"></p>
				<p data-role="success" class="hidden"></p>
				<section data-role="owner-google-setup" class="hidden">
					<p data-role="error" class="hidden"></p>
				</section>
				<section data-role="owner-password-setup">
					<p data-role="error" class="hidden"></p>
					<input type="password">
				</section>
				<button data-role="signin" type="button">
					<span data-role="text">Create Password</span>
				</button>
			</form>
		`,
	});
	const { OwnerSetupForm } = await import("../../src/script/login/forms.mjs");
	const form = document.querySelector("[data-role='owner-setup']");
	const passwordSetup = form.querySelector(
		"[data-role='owner-password-setup']",
	);
	const passwordError = passwordSetup.querySelector("[data-role='error']");
	const ownerSetup = new OwnerSetupForm({}, form);
	ownerSetup.data = {
		email: "owner@example.test",
		error: "Provider unavailable",
	};
	ownerSetup.show();

	assert.equal(
		ownerSetup.error,
		passwordError,
		"Password-only owner setup did not select its visible error slot",
	);
	assert.equal(
		passwordSetup.classList.contains("hidden"),
		false,
		"Password-only owner setup hid its available setup pane",
	);
	assert.equal(
		form.classList.contains("hidden"),
		false,
		"Password-only owner setup did not open",
	);
	assert.equal(
		passwordError.textContent,
		"Provider unavailable",
		"Password-only owner setup did not display its error safely",
	);
});
